lib/datadog/tracing/sampling/rule_sampler.rb

```diff
@@ -29,21 +29,34 @@ def initialize(
           default_sample_rate: Datadog.configuration.tracing.sampling.default_rate,
           default_sampler: nil
         )
-          @rules = rules
-          @rate_limiter = if rate_limiter
+          # AppSec events are sent to the backend using traces.
+          # Standalone ASM billing means that we don't want to charge clients for APM traces,
+          # so we want to send the minimum amount of traces possible (idealy only traces that contains security events),
+          # but for features such as API Security, we need to send at least one trace per minute,
+          # to keep the service alive on the backend side.
+          @rules = if Datadog.configuration.appsec.standalone.enabled
+                     [SimpleRule.new(sample_rate: 1.0)]
+                   elsif default_sample_rate && !default_sampler
+                     # Add to the end of the rule list a rule always matches any trace
+                     rules << SimpleRule.new(sample_rate: default_sample_rate)
+                   else
+                     rules
+                   end
+          @rate_limiter = if Datadog.configuration.appsec.standalone.enabled
+                            # 1 trace per minute
+                            Core::TokenBucket.new(1.0 / 60, 1.0)
+                          elsif rate_limiter
                             rate_limiter
                           elsif rate_limit
                             Core::TokenBucket.new(rate_limit)
                           else
                             Core::UnlimitedLimiter.new
                           end
-
-          @default_sampler = if default_sampler
-                               default_sampler
-                             elsif default_sample_rate
-                               # Add to the end of the rule list a rule always matches any trace
-                               @rules << SimpleRule.new(sample_rate: default_sample_rate)
+          @default_sampler = if Datadog.configuration.appsec.standalone.enabled ||
+              (default_sample_rate && !default_sampler)
```

Comment:
I suggest moving this logic to the initialization of the `RuleSampler`, instead of inside of it:
https://github.com/DataDog/dd-trace-rb/blob/d3d7fde89f4a5bba1032d60867c00ddc669095f7/lib/datadog/tracing/component.rb#L70-L96

It looks like this logic can manipulate the parameters passed to `RuleSampler` to accomplish the same goal, I think it would be: `RuleSampler.new([], rate_limiter: Core::TokenBucket.new(1.0 / 60, 1.0), default_sample_rate: 1.0)`.

Reply:
It is also initialised in here:
https://github.com/DataDog/dd-trace-rb/blob/60febf8012de92758b85a7de0287342475ef0fb0/lib/datadog/tracing/tracer.rb#L50-L62

Do you think I should also move the logic there ? Imo it would mean duplicated code

Reply:
You can ignore this one in `dd-trace-rb/lib/datadog/tracing/tracer.rb`. It's not used anymore, just a relic from when the Tracer class itself was public. You can always assume well go through the `Component` initialization.